src/client/cli/commands/archive.js

```diff
@@ -0,0 +1,120 @@
+      description: 'Filename to output archive tarball to.  If not given archive will be written to stdout.',
+      required: false,
+      default: null
+    }
+  },
+  handler: subcommand((opts: {client: RestClient, queryString: string, output?: ?string}) => {
+    const {client, queryString} = opts
+    let dataFetchPromises = []
+    let outputStream
+    let tarball
+
+    return client.queryStream(queryString)
+      .then(response => new Promise((resolve, reject) => {
+        const {output} = opts
+        const queryStream = response.stream()
+        const outStreamName = output || 'standard output'
+        outputStream = (output == null) ? process.stdout : fs.createWriteStream(output)
+
+        tarball = tar.pack()
+        const gzip = zlib.createGzip()
+        tarball.pipe(gzip).pipe(outputStream)
+
+        let objectIds = []
+        function fetchBatch (force: boolean = false) {
+          if (force || objectIds.length >= OBJECT_BATCH_SIZE) {
+            const ids = objectIds
+            objectIds = []
+            dataFetchPromises.push(writeDataObjectsToTarball(client, tarball, ids))
+          }
+        }
+
+        outputStream.on('error', err => {
+          reject(new Error(`Error writing to ${outStreamName}: ${err.message}`))
+        })
+
+        queryStream.on('error', err => {
+          reject(new Error(`Error reading from query result stream: ${err.message}`))
+        })
+
+        queryStream.on('data', obj => {
+          let stmt
+          try {
+            stmt = Statement.fromProtobuf(obj)
+          } catch (err) {
+            // ignore non-statement results
+            return
+          }
+
+          const name = `stmt/${stmt.id}`
+          const content = Buffer.from(JSON.stringify(obj), 'utf-8')
+          writeToTarball(tarball, name, content)
+
+          for (const id of stmt.objectIds) {
+            objectIds.push(id)
+          }
+          fetchBatch()
+        })
+
+        queryStream.on('end', () => {
+          fetchBatch(true)
+          resolve()
+        })
+      }))
+      .then(() => Promise.all(dataFetchPromises))
+      .then(() => new Promise(resolve => {
+        outputStream.on('end', () => resolve())
+        tarball.finalize()
+      }))
+  })
+}
+
+function writeToTarball (tarball: Object, filename: string, content: Buffer) {
+  const header = Object.assign({}, {name: filename, size: content.length}, TAR_ENTRY_OPTS)
+  tarball.entry(header, content)
+}
+
+function writeDataObjectsToTarball (client: RestClient, tarball: Object, objectIds: Array<string>): Promise<*> {
+  if (objectIds.length < 1) return Promise.resolve()
+
+  return client.batchGetDataStream(objectIds, false)
+    .then(stream => new Promise((resolve, reject) => {
+      stream.on('data', dataResult => {
+        const key = objectIds.shift()
+        if (dataResult == null || typeof dataResult !== 'object' || dataResult.data == null) return
+
+        const bytes = Buffer.from(dataResult.data, 'base64')
```

Comment:
The concat API returns JSON objects with a base64-encoded `data` field.  The line above decodes it so we can write the raw buffer to the tar entry